nixos/modules/services/networking/zeronsd.nix

```diff
@@ -0,0 +1,103 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.zeronsd;
+in {
+  options.services.zeronsd.enable = mkEnableOption (lib.mdDoc "ZeroNSD");
+
+  options.services.zeronsd.servedNetworks = mkOption {
+    default = [];
+    example = [ "a8a2c3c10c1a68de" ];
+    type = types.listOf types.str;
+    description = lib.mdDoc "List of ZeroTier Network IDs to start zeronsd instance for.";
+  };
+
+  options.services.zeronsd.domain = mkOption {
+    default = "home.arpa";
+    type = types.str;
+    description = lib.mdDoc ''
+      Domain under which zerotier records will be available
+    '';
+  };
+
+  options.services.zeronsd.apiTokenFile = mkOption {
+    default = null;
+    type = types.path;
+    description = lib.mdDoc ''
+      Path to a file containing the API Token for ZeroTier Central
+    '';
+  };
+
+  options.services.zeronsd.logLevel = mkOption {
+    default = "info";
+    type = types.str;
+    description = lib.mdDoc ''
+      Log Level (one of off, error, warn, info, debug, trace)
+    '';
+  };
+
+  options.services.zeronsd.wildcardMode = mkOption {
+    default = false;
+    type = types.bool;
+    description = lib.mdDoc ''
+      Enables wildcard mode, where all member names get a wildcard in this format: `*.<name>.<tld>`; this points at the member's IP address(es).
+    '';
+  };
+
+  options.services.zeronsd.package = mkOption {
+    default = pkgs.zeronsd;
+    defaultText = literalExpression "pkgs.zeronsd";
+    type = types.package;
+    description = lib.mdDoc ''
+      ZeroNSD package to use.
+    '';
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services = builtins.listToAttrs (builtins.map (net: {
+      name = "zeronsd-${net}";
+      value = {
+        description = "ZeroNSD (${net})";
+
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" "zerotierone.service" ];
+        wants = [ "network-online.target" ];
+
+        path = [ cfg.package ];
+
+        serviceConfig = let
+          configFile = pkgs.writeText "zeronsd.json" (builtins.toJSON {
+            domain = cfg.domain;
+            log_level = cfg.logLevel;
+            token = cfg.apiTokenFile;
+            wildcard = cfg.wildcardMode;
+          });
+          execPreScript = pkgs.writeShellScript "zeronsd-pre" ''
```

Comment:
why do this, instead of using systemd's StateDirectory and StateDirectoryMode options?

Reply:
Zeronsd needs to talk to the zerotier-one daemon in order to get information about the available networks and their interfaces. Requests to the daemon are authenticated by a token that is saved in that restricted file by the zerotier-one unit.

I didn't want to run this as root as described in the documentation, so this workaround was required.

Reply:
This could also be done with tmpfiles